File: fan_lsqt/julia/mod_lsqt_01.jl
using LinearAlgebra: dot, norm, mul!
using Printf
using SparseArrays
using SpecialFunctions: besselj
using Infiltrator

function gen_idx_xy2ip(Nx, Ny)
    idx_xy2ip = zeros(Int64, Nx, Ny)
    ip = 0
    for i in 1:Nx, j in 1:Ny # the same ordering as in Python
        ip += 1
        idx_xy2ip[i,j] = ip
    end
    return idx_xy2ip
end


function init_H_and_V(Nx, Ny, W)

    idx_xy2ip = gen_idx_xy2ip(Nx, Ny)

    Npoints = Nx * Ny # total number of sites

    row_H = zeros(Int64, Npoints*4 - Nx*2) # the row indices for H
    col_H = zeros(Int64, Npoints*4 - Nx*2) # the column indices for H
    Hij = -ones(ComplexF64, Npoints*4 - Nx*2) # nonzero Hamiltonian elements

    row_V = zeros(Int64, Npoints*2) # row indices for V
    col_V = zeros(Int64, Npoints*2) # column indices for V
    Vij = zeros(ComplexF64, Npoints*2) # nonzero velocity matrix elements

    row_U = 1:Npoints # row and column indices for U
    col_U = 1:Npoints
    Uij = rand(Npoints)*W .- W/2

    count_H = 0
    count_V = 0

    for i in 1:Nx, j in 1:Ny # the same ordering as in Python
        #
        # (0) # get the index of the center site
        idx_center = idx_xy2ip[i,j]
        #
        # (1) consider the left neighbor (periodic boundary)
        ii = i - 1
        if ii <= 0
            ii += Nx # wrap
        end
        idx_left = idx_xy2ip[ii,j]
        #
        count_H += 1    
        row_H[count_H] = idx_center
        col_H[count_H] = idx_left
        #
        count_V += 1
        row_V[count_V] = idx_center
        col_V[count_V] = idx_left
        Vij[count_V] = im
        #
        # (2) consider the right neighbor (periodic boundary)
        ii = i + 1
        if ii >= Nx + 1
            ii -= Nx # wrap
        end
        idx_right = idx_xy2ip[ii,j]
        #
        count_H += 1
        row_H[count_H] = idx_center
        col_H[count_H] = idx_right
        #
        count_V += 1
        row_V[count_V] = idx_center
        col_V[count_V] = idx_right
        Vij[count_V] = -im
        #
        # (3) consider the upper neighbor (open boundary)
        #@printf("\n -- i=%3d j=%3d idx_center=%d\n", i, j, idx_center)
        #@printf("Periodic BC: idx_left=%3d idx_right=%3d\n" , idx_left, idx_right)
        if j < Ny
            idx_up = idx_xy2ip[i,j+1]
            count_H += 1
            row_H[count_H] = idx_center
            col_H[count_H] = idx_up
            #@printf("Open BC: idx_up=%3d\n", idx_up)
        end
        #
        # (4) consider the down neighbor (open boundary)
        if j > 1
            idx_down = idx_xy2ip[i,j-1]
            count_H += 1
            row_H[count_H] = idx_center
            col_H[count_H] = idx_down
            #@printf("Open BC: index_down=%3d\n", idx_down)
        end
    end

    H = sparse( row_H, col_H, Hij )
    U = sparse( row_U, col_U, Uij )
    H = H + U
    V = sparse( row_V, col_V, Vij )

    return H, V

end # function

function init_state(N)
    θ = 2π * rand(N)
    ϕ = exp.(im*θ)
    ϕ ./= norm(ϕ);
    return ϕ
end

function jackson_damping(M)
    m = 1:M
    a = 1.0 / (M + 1.0)
    g = (1.0 .- m * a) .* cos.(pi * m * a)
    g .+= a * sin.(pi * m * a) / tan.(pi * a)
    return g
end

function calc_moments(M, H_scaled, ϕ_l, ϕ_r)
    C = zeros(Float64, M)
    calc_moments!(M, H_scaled, ϕ_l, ϕ_r, C)
    return C
end


function calc_moments!(M, H_scaled, ϕ_l, ϕ_r, C)
    #
    ϕ_0 = copy(ϕ_r)
    C[1] = real(dot(ϕ_l, ϕ_0))
    #
    ϕ_1 = H_scaled * ϕ_0
    C[2] = real(dot(ϕ_l, ϕ_1))
    ϕ_2 = similar(ϕ_l)
    #
    for m in 3:M
        #ϕ_2 = 2 * H_scaled * ϕ_1 - ϕ_0
        mul!(ϕ_2, H_scaled, ϕ_1)
        ϕ_2 *= 2
        ϕ_2 -= ϕ_0
        #        
        C[m] = real(dot(ϕ_l, ϕ_2))
        @views ϕ_0[:] = ϕ_1[:]
        @views ϕ_1[:] = ϕ_2[:]
    end
    return
end


function sum_chebyshev(M, C, E_scaled, E_max, g)
    Ne = length(E_scaled)
    s = zeros(Float64, Ne)
    sum_chebyshev!(M, C, E_scaled, E_max, g, s)
    return s
end

# g is damping factor
# C is modified
# TODO: fuse this with calc_moments ?
function sum_chebyshev!(M, C, E_scaled, E_max, g, s)
    # apply the damping factor
    for i in 1:M
        C[i] = C[i] * g[i]
    end
    Ne = length(E_scaled)     # number of energy points
    T0 = ones(Float64, Ne)
    T1 = copy(E_scaled)
    T2 = zeros(Float64, Ne)
    for i in 1:Ne
        s[i] = C[2] * T1[i]
    end
    for m in 3:M
        for i in 1:Ne
            T2[i] = 2.0 * E_scaled[i] * T1[i] - T0[i]
            T0[i] = T1[i]
            T1[i] = T2[i]
            s[i] += C[m] * T2[i]
        end
    end
    for i in 1:Ne
        s[i] *= 2.0
        s[i] += C[1]
        s[i] *= 2.0 / (pi * sqrt(1.0 - E_scaled[i]^2) * E_max)
    end
    return
end

function calc_dos(M, E_max, E_scaled, H_scaled, phi)
    damping_factor = jackson_damping(M) # get the damping factor
    C = calc_moments(M, H_scaled, phi, phi)
    DOS = sum_chebyshev(M, C, E_scaled, E_max, damping_factor)
    return DOS
end



function calc_dos_fused(M, E_max, E_scaled, H_scaled, phi)

    ϕ_l = phi
    ϕ_r = phi

    C = zeros(Float64, M)
    ϕ_0 = copy(ϕ_r)
    C[1] = real(dot(ϕ_l, ϕ_0))
    #
    ϕ_1 = H_scaled * ϕ_0
    C[2] = real(dot(ϕ_l, ϕ_1))
    #
    for m in 3:M
        ϕ_2 = 2 * H_scaled * ϕ_1 - ϕ_0
        C[m] = real(dot(ϕ_l, ϕ_2))
        @views ϕ_0[:] = ϕ_1[:]
        @views ϕ_1[:] = ϕ_2[:]
    end

    g = jackson_damping(M) # get the damping factor

    # apply the damping factor
    for i in 1:M
        C[i] = C[i] * g[i]
    end
    Ne = length(E_scaled)     # number of energy points
    T0 = ones(Float64, Ne)
    T1 = copy(E_scaled)
    T2 = zeros(Float64, Ne)
    s = zeros(Float64, Ne) # output DOS
    for i in 1:Ne
        s[i] = C[2] * T1[i]
    end
    for m in 3:M
        for i in 1:Ne
            T2[i] = 2.0 * E_scaled[i] * T1[i] - T0[i]
            T0[i] = T1[i]
            T1[i] = T2[i]
            s[i] += C[m] * T2[i]
        end
    end
    for i in 1:Ne
        s[i] *= 2.0
        s[i] += C[1]
        s[i] *= 2.0 / (pi * sqrt(1.0 - E_scaled[i]^2) * E_max)
    end

    return s
end


function evolve( H_scaled, dt_scaled, sign, phi_i )
    j0 = besselj(0, dt_scaled)
    j1 = besselj(1, dt_scaled)
    N = size(phi_i, 1)
    #
    phi_0 = copy(phi_i) # phi_i is not modified
    phi_1 = H_scaled * phi_i
    phi_o = j0*phi_0 + 2.0 * (-1im * sign) * j1 * phi_1
    phi_2 = zeros(ComplexF64, N)
    m = 2
    Hv = similar(phi_1)
    while true
        jm = besselj(m, dt_scaled)
        if abs(jm) < 1.0e-15
            break
        end
        mul!(Hv, H_scaled, phi_1)
        @views phi_2[:] = 2.0 * Hv - phi_0
        @views phi_o[:] .+= 2.0 * (-im*sign)^m * jm * phi_2
        @views phi_0[:] = phi_1[:]
        @views phi_1[:] = phi_2[:]
        m += 1
    end
    return phi_o
end

function calc_vac(M, E_max, dt_scaled, E_scaled, H_scaled, V, ϕ, DOS)
    Ne = length(E_scaled)  # number of energy points
    Nt = length(dt_scaled) # number of time steps
    dt = dt_scaled / E_max
    phi_left = ϕ
    phi_right = V * ϕ
    DOS_times_VAC_old = zeros(Float64, Ne)
    DOS_times_VAC_new = zeros(Float64, Ne)
    VAC = zeros(Float64, Nt, Ne)
    sigma_from_VAC = zeros(Float64, Nt, Ne)
    damping_factor = jackson_damping(M)
    tmp = zeros(Float64, Ne)
    C = zeros(Float64, M)
    Vf = similar(phi_left)
    #
    for nt in 1:Nt
        mul!(Vf, V, phi_left)
        calc_moments!(M, H_scaled, Vf, phi_right, C)
        sum_chebyshev!(M, C, E_scaled, E_max, damping_factor, DOS_times_VAC_new)
        for i in 1:Ne
            VAC[nt,i] = DOS_times_VAC_new[i] / DOS[i]
        end
        if nt > 1
            for i in 1:Ne
                tmp[i] = dt[nt-1] * (DOS_times_VAC_old[i] + DOS_times_VAC_new[i]) * 0.5
                sigma_from_VAC[nt,i] = sigma_from_VAC[nt-1,i] + tmp[i]
            end
        end
        for i in 1:Ne
            DOS_times_VAC_old[i] = DOS_times_VAC_new[i]
        end
        @views phi_left[:] = evolve(H_scaled, dt_scaled[nt], -1, phi_left)
        @views phi_right[:] = evolve(H_scaled, dt_scaled[nt], -1, phi_right)
    end
    sigma_from_VAC *= 2.0 * pi # from e^2/hbar to e^2/h
    return VAC, sigma_from_VAC
end


function evolvex(H_scaled, V_scaled, dt_scaled, phi_i)
    phi_0 = copy(phi_i)
    phix_0 = zeros(ComplexF64, size(phi_i)) #phi_i * 0.0
    phi_1 = H_scaled * phi_0
    phix_1 = im * V_scaled * phi_0
    phi_o = -2im * besselj(1, dt_scaled) * phix_1
    m = 2
    Hv = similar(phi_1)
    Vv = similar(phi_1)
    while true
        jm = besselj(m, dt_scaled)
        if abs(jm) < 1.0e-15
            break
        end
        mul!(Hv, H_scaled, phi_1)
        phi_2 = 2.0 * Hv - phi_0
        mul!(Vv, V_scaled, phi_1)
        phix_2 = 2im * Vv
        #
        mul!(Hv, H_scaled, phix_1)
        phix_2 += 2.0 * Hv - phix_0
        #
        @views phi_o[:] .+= 2.0 * (-1im)^m * jm * phix_2[:]
        #
        @views phi_0[:] = phi_1[:] 
        @views phi_1[:] = phi_2[:]
        #
        @views phix_0[:] = phix_1[:]
        @views phix_1[:] = phix_2[:]
        m += 1
    end
    return phi_o
end


function calc_msd(M, E_max, dt_scaled, E_scaled, H_scaled, V_scaled, phi, DOS)
    Ne = length(E_scaled)  # number of energy points
    Nt = length(dt_scaled) # number of time steps
    dt = dt_scaled / E_max
    MSD = zeros((Nt, Ne))
    sigma_from_MSD = zeros((Nt, Ne))
    phix = zeros(size(phi))
    DOS_times_MSD_old = zeros(Ne)
    DOS_times_MSD_new = zeros(Ne)
    damping_factor = jackson_damping(M)
    C = zeros(Float64, M)
    for nt in 1:Nt
        #
        phix = evolve(H_scaled, dt_scaled[nt], 1, phix)
        phix += evolvex(H_scaled, V_scaled, dt_scaled[nt], phi)
        phi = evolve(H_scaled, dt_scaled[nt], 1, phi)
        #
        calc_moments!(M, H_scaled, phix, phix, C)
        sum_chebyshev!(M, C, E_scaled, E_max, damping_factor, DOS_times_MSD_new)
        #
        @views MSD[nt,:] = DOS_times_MSD_new ./ DOS
        @views sigma_from_MSD[nt,:] = (DOS_times_MSD_new - DOS_times_MSD_old) / dt[nt] * 0.5
        @views DOS_times_MSD_old[:] = DOS_times_MSD_new[:]
    end
    sigma_from_MSD *= 2π # from e^2/hbar to e^2/h
    return MSD, sigma_from_MSD
end

File: fan_lsqt/python/mod_lsqt_01.py
import numpy as np
import matplotlib.pyplot as plt
from scipy import sparse
from scipy import special

def find_index(nx, ny, Ny):
    index = nx * Ny + ny
    return index

def find_H(Nx, Ny, W):
    N = Nx * Ny; # total number of sites
    row_H = np.zeros(N * 4 - Nx * 2) # the row indices for H
    col_H = np.zeros(N * 4 - Nx * 2) # the column indices for H
    Hij = -np.ones(N * 4 - Nx * 2, complex) # nonzero Hamiltonian elements
    row_V = np.zeros(N * 2) # row indices for V
    col_V = np.zeros(N * 2) # column indices for V
    Vij = np.zeros(N * 2, complex) # nonzero velocity matrix elements
    row_U = np.arange(0, N) # row and column indices for U
    Uij = np.random.uniform(-W * 0.5, W * 0.5, N) # on-site potentials
    count_H = 0 # number of nonzero H elements
    count_V = 0 # number of nonzero V elements
    for nx in range(Nx):
        for ny in range(Ny):
            # (0) # get the index of the center site
            index_center = find_index(nx, ny, Ny)
            # (1) consider the left neighbor (periodic boundary)
            index_left = find_index((nx - 1) % Nx, ny, Ny)
            row_H[count_H] = index_center
            col_H[count_H] = index_left
            count_H += 1
            row_V[count_V] = index_center
            col_V[count_V] = index_left
            Vij[count_V] = 1j
            count_V += 1
            # (2) consider the right neighbor (periodic boundary)
            index_right = find_index((nx + 1) % Nx, ny, Ny)
            row_H[count_H] = index_center
            col_H[count_H] = index_right
            count_H += 1
            row_V[count_V] = index_center
            col_V[count_V] = index_right
            Vij[count_V] = -1j
            count_V += 1
            # (3) consider the upper neighbor (open boundary)
            if ny < Ny - 1:
                index_up = find_index(nx, (ny + 1), Ny)
                row_H[count_H] = index_center
                col_H[count_H] = index_up
                count_H += 1
            # (4) consider the down neighbor (open boundary)
            if ny > 0:
                index_down = find_index(nx, (ny - 1), Ny)
                row_H[count_H] = index_center
                col_H[count_H] = index_down
                count_H += 1
    H = sparse.csr_matrix((Hij, (row_H, col_H)), shape = (N, N))
    U = sparse.csr_matrix((Uij, (row_U, row_U)), shape = (N, N))
    H = H + U
    V = sparse.csr_matrix((Vij, (row_V, col_V)), shape = (N, N))
    return (H, V)

def create_state(N):
    random_phase = np.random.uniform(0, 2 * np.pi, N)
    phi = np.cos(random_phase) + np.sin(random_phase) * 1j; 
    phi /= np.linalg.norm(phi);
    return phi


def find_moments(M, H_scaled, phi_left, phi_right):  
    C = np.zeros(M)
    phi_0 = phi_right
    C[0] = np.vdot(phi_left, phi_0).real
    phi_1 = H_scaled.dot(phi_0)
    C[1] = np.vdot(phi_left, phi_1).real
    for m in range(2, M):
        phi_2 = 2.0 * H_scaled.dot(phi_1) - phi_0
        C[m] = np.vdot(phi_left, phi_2).real
        phi_0 = phi_1
        phi_1 = phi_2
    return C


def jackson_damping(M):
    m = np.arange(M)
    a = 1.0 / (M + 1.0)
    g = (1.0 - m * a) * np.cos(np.pi * m * a)
    g += a * np.sin(np.pi * m * a) / np.tan(np.pi * a)
    return g


def plot_jackson_damping():
    M = 10000
    plt.plot(np.arange(M), jackson_damping(M))
    plt.xlabel('m')
    plt.ylabel('g')
    plt.show()


def chebyshev_summation(M, C, E_scaled, E_max):
    g = jackson_damping(M) # get the damping factor
    C *= g                 # apply the damping factor
    Ne = E_scaled.size     # number of energy points
    T0 = np.ones(Ne)
    T1 = E_scaled
    s = C[1] * T1
    for m in range(2, M):
        T2 = 2.0 * E_scaled * T1 - T0
        T0 = T1
        T1 = T2
        s += C[m] * T2
    s *= 2.0
    s += C[0] * np.ones(Ne)
    s *= 2.0 / (np.pi * np.sqrt(1.0 - E_scaled * E_scaled) * E_max)
    return s


def find_dos(M, E_max, E_scaled, H_scaled, phi):
    C = find_moments(M, H_scaled, phi, phi)
    DOS = chebyshev_summation(M, C, E_scaled, E_max)
    return DOS

def evolve(H_scaled, dt_scaled, sign, phi_i):
    phi_0 = phi_i
    phi_1 = H_scaled.dot(phi_i)
    j0 = special.jv(0, dt_scaled)
    j1 = special.jv(1, dt_scaled)
    phi_o = j0 * phi_0 + 2.0 * (-1j * sign) * j1 * phi_1
    m = 2
    while True:
        jm = special.jv(m, dt_scaled)
        if abs(jm) < 1.0e-15:
            break
        phi_2 = 2.0 * H_scaled.dot(phi_1) - phi_0
        phi_o += 2.0 * (-1j * sign) ** m * jm * phi_2
        phi_0 = phi_1
        phi_1 = phi_2
        m += 1
    return phi_o


def evolvex(H_scaled, V_scaled, dt_scaled, phi_i):
    phi_0 = phi_i
    phix_0 = phi_i * 0.0
    phi_1 = H_scaled.dot(phi_0)
    phix_1 = 1j * V_scaled.dot(phi_0)
    phi_o = (-2.0j) * special.jv(1, dt_scaled) * phix_1
    m = 2
    while True:
        jm = special.jv(m, dt_scaled)
        if abs(jm) < 1.0e-15:
            break
        phi_2 = 2.0 * H_scaled.dot(phi_1) - phi_0
        phix_2 = 2.0j * V_scaled.dot(phi_1)
        phix_2 += 2.0 * H_scaled.dot(phix_1) - phix_0
        phi_o += 2.0 * (-1j) ** m * jm * phix_2
        phi_0 = phi_1 
        phi_1 = phi_2
        phix_0 = phix_1
        phix_1 = phix_2
        m += 1
    return phi_o


def find_vac(M, E_max, dt_scaled, E_scaled, H_scaled, V, phi, DOS):
    Ne = E_scaled.size  # number of energy points
    Nt = dt_scaled.size # number of time steps
    dt = dt_scaled / E_max
    phi_left = phi
    phi_right = V.dot(phi)
    DOS_times_VAC_old = np.zeros(Ne)
    DOS_times_VAC_new = np.zeros(Ne)
    VAC = np.zeros((Nt, Ne))
    sigma_from_VAC = np.zeros((Nt, Ne))
    for nt in range(Nt):
        C = find_moments(M, H_scaled, V.dot(phi_left), phi_right)
        DOS_times_VAC_new = chebyshev_summation(M, C, E_scaled, E_max)
        VAC[nt, :] = DOS_times_VAC_new / DOS
        if nt > 0:
            tmp = dt[nt - 1] * (DOS_times_VAC_old + DOS_times_VAC_new) * 0.5
            sigma_from_VAC[nt, :] = sigma_from_VAC[nt - 1, :] + tmp
        DOS_times_VAC_old = DOS_times_VAC_new
        phi_left = evolve(H_scaled, dt_scaled[nt], -1, phi_left)
        phi_right = evolve(H_scaled, dt_scaled[nt], -1, phi_right)
    sigma_from_VAC *= 2.0 * np.pi # from e^2/hbar to e^2/h
    return VAC, sigma_from_VAC



def find_msd(M, E_max, dt_scaled, E_scaled, H_scaled, V_scaled, phi, DOS):
    Ne = E_scaled.size  # number of energy points
    Nt = dt_scaled.size # number of time steps
    dt = dt_scaled / E_max
    MSD = np.zeros((Nt, Ne))
    sigma_from_MSD = np.zeros((Nt, Ne))
    phix = phi * 0.0
    DOS_times_MSD_old = np.zeros(Ne)
    DOS_times_MSD_new = np.zeros(Ne)
    for nt in range(Nt):
        phix = evolve(H_scaled, dt_scaled[nt], 1, phix);
        phix += evolvex(H_scaled, V_scaled, dt_scaled[nt], phi);
        phi = evolve(H_scaled, dt_scaled[nt], 1, phi);
        C = find_moments(M, H_scaled, phix, phix);
        DOS_times_MSD_new = chebyshev_summation(M, C, E_scaled, E_max);
        MSD[nt, :] = DOS_times_MSD_new / DOS
        sigma_from_MSD[nt, :] = (DOS_times_MSD_new - DOS_times_MSD_old) / dt[nt] * 0.5
        DOS_times_MSD_old = DOS_times_MSD_new
    sigma_from_MSD *= 2.0 * np.pi # from e^2/hbar to e^2/h
    return MSD, sigma_from_MSD



def lsqt(Nx, Ny, W, M, E_max, E, dt):
    
    print("Find H ...", end=" ")
    H, V = find_H(Nx, Ny, W)
    print("... end of find H")
    
    phi = create_state(Nx * Ny)
    
    print("Find DOS ...", end=" ")
    DOS = find_dos(M, E_max, E/E_max, H/E_max, phi)
    print("... end of find DOS")

    print("Find VAC ... ", end=" ")
    VAC, sigma_from_VAC = find_vac(M, E_max, dt*E_max, E/E_max, H/E_max, V, phi, DOS)
    print("... end of find VAC")

    print("Finding MSD ... ", end=" ")
    MSD, sigma_from_MSD = find_msd(M, E_max, dt*E_max, E/E_max, H/E_max, V/E_max, phi, DOS)
    print("... end of find MSD")

    return DOS, VAC, sigma_from_VAC, MSD, sigma_from_MSD